README.md
# @saber2pr/jsx-ast-parser

[![npm](https://img.shields.io/npm/v/@saber2pr/jsx-ast-parser.svg?color=blue)](https://www.npmjs.com/package/@saber2pr/jsx-ast-parser)
[![codecov](https://codecov.io/gh/Saber2pr/jsx-ast-parser/branch/master/graph/badge.svg?token=DI9E88OIZU)](https://codecov.io/gh/Saber2pr/jsx-ast-parser)

> jsx parser by parser combinators.

```bash
yarn add @saber2pr/jsx-ast-parser
```

### Usage

```ts
import {
  parser,
  transformer,
  compiler,
  traverser,
} from '@saber2pr/jsx-ast-parser'

const code = `<div className="hello">world</div>`

const ast = parser.parse(code) // parse ast from code string
const jsx = transformer.transform(ast) // transform ast to jsx

// compile jsx to source code
compiler.compile(jsx) === code // true

// find jsx node
traverser.findNode(jsx, node => transformer.isTextElement(node)) // [ { tagName: 'text', nodeValue: 'world' } ]
```

#### Help

1. [Api Docs](https://saber2pr.top/jsx-ast-parser/)
2. [Jsx Ast Viewer](https://jsx-ast-viewer.vercel.app/)

- Profile
  - [see parser output ast.json](./public/ast.json)
  - [see transformer output jsx.json](./public/jsx.json)
  - [see compiler output out.jsx](./public/out.jsx)

### Feature

#### Overview Grammar BNF

[jsx.bnf](./public/jsx.bnf)

#### Syntax currently supported

- [ ] parser
  - [ ] jsx
    - [x] jsx opened
    - [x] jsx self closing
    - [ ] jsx props
      - [x] jsx props string-value
      - [x] jsx props number-value, bool-value
      - [x] jsx props object
      - [x] jsx props object-array
      - [x] jsx props string-array, number-array
      - [x] jsx props arrow function
        - [x] jsx props arrow function scope statements
      - [x] jsx props function
        - [x] jsx props function scope statements
  - [ ] statement
    - [x] call chain
    - [x] arrow function
    - [x] comment
    - [ ] arithmetic
    - [x] function
    - [x] define
    - [x] assign
    - [x] if else
    - [ ] for
    - [ ] while
    - [ ] try catch
    - [x] return
    - [ ] deconstruct
    - [ ] as
- [ ] transformer
  - [x] transform jsx
- [ ] traverser
  - [ ] traverse ast
  - [x] traverse jsx
- [ ] compiler
  - [ ] compile ast
  - [x] compile jsx

### Why

It started as a project for me to learn the principles of compilation, but now I'm confident I can make it better! I will continue to provide analysis tools for JSX code.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.2](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.1.1...v0.1.2) (2021-10-08)


### Features

* arrow function args ([ab4b82b](https://github.com/Saber2pr/jsx-ast-parser/commit/ab4b82b8082380c0b150fd90b2bda3328007b5b0))
* jsx inner expr ([f8f2c10](https://github.com/Saber2pr/jsx-ast-parser/commit/f8f2c104f942b1d9a492aea04c3d97be9294cda1))

### [0.1.1](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.0.9...v0.1.1) (2021-10-07)


### Features

* tests ([55c998e](https://github.com/Saber2pr/jsx-ast-parser/commit/55c998e295b511efe8d580710420d967503b059f))

### [0.0.9](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.0.8...v0.0.9) (2021-10-06)


### Features

* statement if else ([2e6b007](https://github.com/Saber2pr/jsx-ast-parser/commit/2e6b007274196e27df756aa0d28c637bcb5c445b))
* statement return ([c86b4ca](https://github.com/Saber2pr/jsx-ast-parser/commit/c86b4ca3606050935c391ec6d8ac27fbc0fd7253))

### [0.0.8](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.0.7...v0.0.8) (2021-10-05)


### Features

* bnf ([5654ecf](https://github.com/Saber2pr/jsx-ast-parser/commit/5654ecfe0d984427e609c1942deda397984abb39))

### [0.0.7](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.0.6...v0.0.7) (2021-10-04)


### Features

* if statement ([356b7d7](https://github.com/Saber2pr/jsx-ast-parser/commit/356b7d7a640f335882a2c9e2581a27e67542342a))

### [0.0.6](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.0.5...v0.0.6) (2021-10-04)


### Features

* comment ([2bae7ad](https://github.com/Saber2pr/jsx-ast-parser/commit/2bae7ad155b42504af768bcb3728447a371931a7))
* statement define, assign ([60b2152](https://github.com/Saber2pr/jsx-ast-parser/commit/60b2152b4d469d57b2ab808c634530d29ae9a1e6))
* tests coverage ci ([332b1c1](https://github.com/Saber2pr/jsx-ast-parser/commit/332b1c1c5dc45d0aab26c22eea9ec2f792093cf8))

### [0.0.5](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.0.4...v0.0.5) (2021-10-03)


### Features

* jsx props function ([2ea4c25](https://github.com/Saber2pr/jsx-ast-parser/commit/2ea4c25a2d7ebfc375feeca1e1e2a5d962874b01))
* transformExpression ([3c1274c](https://github.com/Saber2pr/jsx-ast-parser/commit/3c1274c7356b79f5948580e4b14073dc41bf6619))

### [0.0.4](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.0.3...v0.0.4) (2021-10-03)


### Features

* transform factory ([0dc8204](https://github.com/Saber2pr/jsx-ast-parser/commit/0dc8204856945d68e40fcfd276a8507ef6d0e7da))
* traverse arrow function & callchain ([96ec78d](https://github.com/Saber2pr/jsx-ast-parser/commit/96ec78ddc6c4559f2175b70c80b72d6d996eecd6))

### [0.0.3](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.0.2...v0.0.3) (2021-10-02)


### Features

* ArrowFunctionExpr & CallChainExpr ([a86f0f9](https://github.com/Saber2pr/jsx-ast-parser/commit/a86f0f95a67a34517f41c17a6b59fabdd4bf7c0c))
* compile arrow function & callchain ([34b079e](https://github.com/Saber2pr/jsx-ast-parser/commit/34b079eca08ac4b7f33842c8f36cd88344185af7))
* jsx props identity ([b710d52](https://github.com/Saber2pr/jsx-ast-parser/commit/b710d5224078f47d76eb9e38424664eee669f5c5))
* tests arrow function & callchain ([99bf0d4](https://github.com/Saber2pr/jsx-ast-parser/commit/99bf0d4cbf55f855e7c7bd5fdf72aa1d9f10b5b9))

### [0.0.2](https://github.com/Saber2pr/jsx-ast-parser/compare/v0.0.1...v0.0.2) (2021-10-02)


### Features

* traverser ([6cdea89](https://github.com/Saber2pr/jsx-ast-parser/commit/6cdea89f33ca05a922dedc93167e991b221a8599))

### 0.0.1 (2021-10-02)


### Features

* ast.type ([466bf53](https://github.com/Saber2pr/jsx-ast-parser/commit/466bf5344134c3644ae111c3e39370326bb5c86d))
* compiler ([10d3db4](https://github.com/Saber2pr/jsx-ast-parser/commit/10d3db4b1e65692bb7771c4f51a19c29df0528a4))
* init ([e627ca5](https://github.com/Saber2pr/jsx-ast-parser/commit/e627ca5ea51a229688b13c159ab784661ec80ba0))
* jsx props number-value, bool-value ([8a3a048](https://github.com/Saber2pr/jsx-ast-parser/commit/8a3a0482e0b325d8c695a7dfbb0f494e14edda20))
* jsx props object ([8b372b4](https://github.com/Saber2pr/jsx-ast-parser/commit/8b372b45c2c5643a2eedabdeba8dc51ac1644cd4))
* test ci ([b35a90f](https://github.com/Saber2pr/jsx-ast-parser/commit/b35a90fc9b7f862f4da44f0351d7db889a377608))
* transform LL1 tests ([3f13f9b](https://github.com/Saber2pr/jsx-ast-parser/commit/3f13f9bbc75fe57681ebbc6911cc9993bbcf3bd5))
* transformer ([da0d6b7](https://github.com/Saber2pr/jsx-ast-parser/commit/da0d6b754cda345f1770333320fc8602713f0522))


### Bug Fixes

* transformPropsExpr ([614d4a3](https://github.com/Saber2pr/jsx-ast-parser/commit/614d4a3935256c758fddf645846f5d90210af679))


src/parser/Consumer.ts
/*
 * @Author: saber2pr
 * @Date: 2021-09-12 12:06:27
 * @Last Modified by: saber2pr
 * @Last Modified time: 2021-10-08 19:38:57
 */
import * as parsec from 'typescript-parsec'

import * as Ast from './Ast'
import { TokenKind } from './Tokenizer'

export type Token = parsec.Token<TokenKind>

// basic

export function applyNumber(token: Token): Ast.NumberExpr {
  return {
    kind: 'NumberExpr',
    value: +token.text,
  }
}

export function applyString(text: Ast.TextExpr): Ast.StringExpr {
  return {
    kind: 'StringExpr',
    value: text.value,
  }
}

export function applyKeyword(token: Token): Ast.KeywordExpr {
  return {
    kind: 'KeywordExpr',
    name: token.text,
  }
}

export function applyIdentity(
  source: [Token, [Ast.NumberExpr, Ast.TextExpr] | undefined]
): Ast.IdentityExpr {
  const [letter, tail] = source
  let name = letter.text
  if (tail) {
    const [digit, text] = tail
    if (digit) {
      name += digit.value
    }
    if (text) {
      name += text.value
    }
  }
  return {
    kind: 'IdentityExpr',
    name,
  }
}

// Jsx

export function applyProp(
  source:
    | [Ast.IdentityExpr, Ast.PropExpr['value']]
    | [Ast.IdentityExpr, undefined]
): Ast.PropExpr {
  const [name, token] = source
  let value: Ast.PropExpr['value']
  if (token) {
    value = token
  } else {
    value = {
      kind: 'JsxInnerExpr',
      body: {
        kind: 'IdentityExpr',
        name: 'true',
      } as Ast.IdentityExpr,
    }
  }
  return {
    kind: 'PropExpr',
    key: name,
    value,
  }
}

export function applyObject(
  source:
    | [Ast.IdentityExpr, Token, Ast.Expression | Ast.IdentityExpr][]
    | undefined = []
): Ast.ObjectExpr {
  return {
    kind: 'ObjectExpr',
    props: source.reduce((acc, cur) => ({ ...acc, [cur[0].name]: cur[2] }), {}),
  }
}

export function applyArray(
  items: (Ast.Expression | Ast.IdentityExpr)[] | undefined = []
): Ast.ArrayExpr {
  return {
    kind: 'ArrayExpr',
    items: items ?? [],
  }
}

export function applyOpeningTag(
  source: [Ast.IdentityExpr, Ast.PropExpr[]]
): Ast.OpeningTagExpr {
  const [name, value] = source
  return {
    kind: 'OpeningTagExpr',
    tagName: name,
    props: value,
  }
}

export function applyClosingTag(source: Ast.IdentityExpr): Ast.ClosingTagExpr {
  return {
    kind: 'ClosingTagExpr',
    tagName: source,
  }
}

export function applyJsxSelfClosing(
  source: [Ast.IdentityExpr, Ast.PropExpr[]]
): Ast.JsxSelfClosingExpr {
  const [name, value] = source
  return {
    kind: 'JsxSelfClosingExpr',
    tagName: name,
    props: value,
  }
}

export function applyText(source: Token[]): Ast.TextExpr {
  return {
    kind: 'TextExpr',
    value: source.map(token => token.text).join(''),
  }
}

export function applyJsx(
  source: [Ast.OpeningTagExpr, Ast.JsxExpr['body'], Ast.ClosingTagExpr]
): Ast.JsxExpr {
  return {
    kind: 'JsxExpr',
    openingTag: source[0],
    body: source[1],
    closingTag: source[2],
  }
}

export function applyJsxInner(
  source: Ast.Expression | Ast.IdentityExpr
): Ast.JsxInnerExpr {
  return {
    kind: 'JsxInnerExpr',
    body: source,
  }
}

// Statement

export function applyArrowFunction(
  source: [Ast.Parameter | Ast.IdentityExpr, Ast.Expression]
): Ast.ArrowFunctionExpr {
  const [args = [], body] = source
  return {
    kind: 'ArrowFunctionExpr',
    args,
    body,
  }
}

export function applyFunction(
  source: [Ast.IdentityExpr | undefined, Ast.Parameter, Ast.BlockExpr]
): Ast.FunctionExpr {
  const [name, args = [], body] = source
  return {
    kind: 'FunctionExpr',
    name,
    args,
    body,
  }
}

export function applyCallChain(
  source: [Ast.IdentityExpr[], Ast.Parameter]
): Ast.CallChainExpr {
  const [chain, args = []] = source
  return {
    kind: 'CallChainExpr',
    caller: chain[0],
    chain: chain.slice(1),
    args: args,
  }
}

export function applyVariableAssign(
  source: [Ast.IdentityExpr, Ast.Expression | undefined]
): Ast.VariableAssignExpr {
  const [name, value] = source
  return {
    kind: 'VariableAssignExpr',
    name,
    value,
  }
}

export function applyBlock(body: Ast.Statement[] = []): Ast.BlockExpr {
  return {
    kind: 'BlockExpr',
    body,
  }
}

// statement

export function applyDefineVariable(
  source: [Ast.KeywordExpr, Ast.VariableAssignExpr | Ast.IdentityExpr]
): Ast.DefineVariableStatement {
  const [type, assign] = source
  return {
    kind: 'DefineVariableStatement',
    type,
    assign,
  }
}

export function applyIf(
  source: [
    Ast.IfStatement['args'],
    Ast.IfStatement['body'],
    Ast.IfStatement['els']
  ]
): Ast.IfStatement {
  const [args = [], body, els] = source
  return {
    kind: 'IfStatement',
    args,
    body,
    els,
  }
}

export function applyReturn(source: Ast.Expression): Ast.ReturnStatement {
  return {
    kind: 'ReturnStatement',
    value: source,
  }
}

export function applyProgram(value: Ast.Program['body']): Ast.Program {
  return {
    kind: 'Program',
    body: value,
  }
}
